### PR TITLE
Merge common example infra

### LINKED ISSUE #924: Merge example duplicated code
Merge the examples duplicated code. Use generics s.t. we can extend this to multiple variations of types.

### PR DESCRIPTION
**Pull request by DieracDelta**
This PR merges common infrastructure into a trait, `CliConfig`, and implements it for both LIbp2p and the Centralizedserver in `examples/infra/mod.rs`. Note: there's a lot of generics. Hopefully as we continue through the refactor, these generic bounds are simplified. Ideally at the end we just have one trait bound that propagates all our generic needs. I've also hardcoded this to ValidatingConsensus. We'll plan to relax this bound later on.

Features included:
- Generate libp2p layer ed25519 keys from an index, removing the need for deploy/keys.
- We can initialize hotshot and execute consensus using exclusively trait functions in the default `CliConfig` trait definition
- Simplify the Cli since we require an orchestrator
- Reduce duplicated information between config structs
- Make communicationchannel structs depend on unused generics. This makes it harder to misuse communicationchannels by including incorrect generics and getting difficult to parse trait bound errors.
- Add some convenience functions to testable structs
- Padding transactions now is implemented for libp2p

Features to be handled in a future PR:
- We need to know all keys ahead of time for libp2p. We should special case a "everyone" topic. This will fix the nll_todos in the libp2p code.
- Document how to run
- Test with orchestrator locally, and add script to do so.
- Nuke `Run` struct
- A sequencing demo that makes sense for both forms of consensus

Closes #924 , part of #923

### CODE REVIEW

File: src/traits/networking/centralized_server_network.rs [hunk 1]
```diff
@@ -993,14 +994,19 @@ impl From<hotshot_centralized_server::Error> for Error {
...
             async_sleep(Duration::from_secs(1)).await;
         }
         true
     }
 
+    #[instrument(name = "CentralizedServer::ready", skip_all)]
+    async fn ready_nonblocking(&self) -> bool {
+        self.run_ready()
+    }
+
```
[the remark this answers is not in the document]
Comment (DieracDelta) [reply src/traits/networking/centralized_server_network.rs:1009]: Yes, these are great suggestions! I'll make the requisite changes. For bullet point 2, why `run_view`? I don't think I modified that function.

[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/traits/networking/centralized_server_network.rs:1009]: Oops, I meant `run_ready`.

File: src/traits/networking/centralized_server_network.rs [hunk 2]
```diff
@@ -1064,16 +1070,28 @@ impl<M: NetworkMsg, K: SignatureKey + 'static, E: ElectionConfig + 'static> Conn
 
 /// libp2p identity communication channel
 #[derive(Clone)]
-pub struct CentralizedCommChannel<TYPES: NodeType>(
+pub struct CentralizedCommChannel<
+    TYPES: NodeType,
+    LEAF: LeafType<NodeType = TYPES>,
+    PROPOSAL: ProposalType<NodeType = TYPES>,
+    ELECTION: Election<TYPES>,
```
[the remark this answers is not in the document]
Comment (DieracDelta) [reply src/traits/networking/centralized_server_network.rs:1077]: Yes, these generics are unused. I added them because I think we don't want the same Communication Channel to be used for multiple different Proposals/Leaf/Elections. So if we constrain these here, this will make it more difficult to misuse the communication channel. 

An example might help portray what I mean: suppose we've implemented CommunicationChannel for CentralizedCommChannel for both VRF style election and static committees. Then, we accidentally invoke `CommunicationChannel` with VRF style election in one place, then in a separate place we invoke `CommunicationChannel` with static committee election. We meant to use static committee in both places.

Constraining the generics on the struct itself prevents the possibility of that sort of bug, which is why I added it. Not sure if it is worth the extra generics though.

File: examples/infra/mod.rs [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+
+pub const LIBP2P_BOOTSTRAPS_REMOTE_IPS: &[&str] = &[
+    "0.ap-south-1.cluster.aws.espresso.network:9000",
+    "1.ap-south-1.cluster.aws.espresso.network:9000",
+    "0.us-east-2.cluster.aws.espresso.network:9000",
+    "1.us-east-2.cluster.aws.espresso.network:9000",
+    "2.us-east-2.cluster.aws.espresso.network:9000",
+    "0.us-west-2.cluster.aws.espresso.network:9000",
+    "1.us-west-2.cluster.aws.espresso.network:9000",
+];
```
Comment (elliedavidson) [inline examples/infra/mod.rs:106]: I think we should put these in a configuration file.  The remote IPs particularly are very likely to change, and it is nice to not have to recompile everything to update them.  My two cents 😄

File: examples/infra/mod.rs [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+        // This assumes all submitted transactions make it through consensus:
+        error!(
+            "{} total bytes submitted in {:?}",
+            total_size, total_time_elapsed
+        );
+        debug!("All rounds completed");
+
+        RunResults {
+            // FIXME nuke this field since we're not doing this anymore.
+            run: Run(0),
```
Comment (elliedavidson) [inline examples/infra/mod.rs:600]: Why aren't we using `run` anymore?  I think I missed that conversation.

File: examples/infra/mod.rs (hunk 3 above)
Comment (elliedavidson) [reply examples/infra/mod.rs:106]: Though I could also work on doing this when I integrate the new orchestrator since I'll need to update the config stuff anyway!

Comment (DieracDelta) [reply examples/infra/mod.rs:106]: Yes, I agree! I'll try to do that in this PR (I think this should be a minor change)

File: src/traits/networking/centralized_server_network.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/traits/networking/centralized_server_network.rs:1077]: Yeah, this makes sense! Let's keep the extra generics then.

File: examples/infra/mod.rs (hunk 3 above)
Comment (DieracDelta) [reply examples/infra/mod.rs:106]: edit: perhaps not so nontrivial. [Added a separate issue for this](https://github.com/EspressoSystems/HotShot/issues/950), I'll tackle it in a separate PR.

File: src/traits/networking/centralized_server_network.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DieracDelta) [reply src/traits/networking/centralized_server_network.rs]: Ended up renaming `ready_blocking` to `wait_for_ready`. Maybe that's too wordy?

File: examples/infra/mod.rs (hunk 4 above)
Comment (elliedavidson) [reply examples/infra/mod.rs:600]: Yes, this sounds perfectly fine!  Thank you for the explanation!

File: src/traits/networking/centralized_server_network.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/traits/networking/centralized_server_network.rs]: `wait_for_ready` is good!

**Review by shenkeyao — APPROVED** [review]
LGTM!
